leakcanary-android-leak-deobfuscation-plugin/src/main/java/com/squareup/leakcanary/leak/deobfuscation/plugin/LeakcanaryLeakDeobfuscationPlugin.kt

```diff
@@ -0,0 +1,109 @@
+
+    val androidAppPluginHandler = { _: Plugin<*> ->
+      val appExtension = project.extensions.getByType(AppExtension::class.java)
+      appExtension.applicationVariants.all { variant ->
+        setupPlugin(project, variant, leakCanaryExtension)
+      }
+    }
+
+    val androidLibraryPluginHandler = { _: Plugin<*> ->
+      val libraryExtension = project.extensions.getByType(LibraryExtension::class.java)
+      libraryExtension.libraryVariants.all { variant ->
+        setupPlugin(project, variant, leakCanaryExtension)
+      }
+    }
+
+    when {
+      project.plugins.hasPlugin("com.android.application") -> {
+        project.plugins.withId("com.android.application", androidAppPluginHandler)
+      }
+      project.plugins.hasPlugin("com.android.library") -> {
+        project.plugins.withId("com.android.library", androidLibraryPluginHandler)
+      }
+      else -> {
+        throw GradleException(
+            "This plugin can only be applied in Android application or library modules."
+        )
+      }
+    }
+  }
+
+  private fun setupPlugin(
+    project: Project,
+    variant: BaseVariant,
+    leakCanaryExtension: LeakCanaryDeobfuscationExtension
+  ) {
+    if (leakCanaryExtension.filterObfuscatedVariants(variant)) {
+      val copyObfuscationMappingFileTaskProvider = project.tasks.register(
+          "leakCanaryCopyObfuscationMappingFor${variant.name.capitalize()}",
+          CopyObfuscationMappingFileTask::class.java
+      ) {
+        it.variantDirName = variant.dirName
+        it.mappingFile = variant.mappingFile
+        it.mergeAssetsDirectory = variant.mergeAssetsProvider.get()
+            .outputDir.get()
+            .asFile
+      }
+
+      val mappingGeneratingTaskProvider =
+        findTaskProviderOrNull(
+            project,
+            "transformClassesAndResourcesWithR8For${variant.name.capitalize()}"
+        ) ?: findTaskProviderOrNull(
+            project,
+            "transformClassesAndResourcesWithProguardFor${variant.name.capitalize()}"
+        )
+
+      getPackageTaskProvider(variant)?.configure {
+        it.dependsOn(copyObfuscationMappingFileTaskProvider)
+      }
+      copyObfuscationMappingFileTaskProvider.configure { copyProguardMappingFileTask ->
+        mappingGeneratingTaskProvider?.let { copyProguardMappingFileTask.dependsOn(it) }
+      }
+    }
+  }
+
+  private fun findTaskProviderOrNull(
+    project: Project,
+    taskName: String
+  ): TaskProvider<Task>? {
+    return try {
+      project.tasks.named(taskName)
+    } catch (proguardTaskNotFoundException: UnknownTaskException) {
+      null
+    }
+  }
+
+  private fun getPackageTaskProvider(variant: BaseVariant): TaskProvider<out DefaultTask>? {
+    return when (variant) {
+      is LibraryVariant -> {
+        variant.packageLibraryProvider
+      }
+      is ApplicationVariant -> {
+        variant.packageApplicationProvider
+      }
+      else -> {
+        null
```

Comment:
it would be better to test at the beginning of the plugin if the android plugin is present, one of the 2, and throw an error if not. You could then also throw an error here.

Reply:
You already filter, you should throw an exception here.